lib/domain/service/file.dart

```diff
@@ -0,0 +1,146 @@
+// Copyright © 2022-2023 IT ENGINEERING MANAGEMENT INC,
+//                       <https://github.com/team113>
+//
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the GNU Affero General Public License v3.0 as published by the
+// Free Software Foundation, either version 3 of the License, or (at your
+// option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License v3.0 for
+// more details.
+//
+// You should have received a copy of the GNU Affero General Public License v3.0
+// along with this program. If not, see
+// <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+// Add hive storing all downloaded files? // next pulls?
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:dio/dio.dart';
+import 'package:path_provider/path_provider.dart';
+
+import '/config.dart';
+import '/util/backoff.dart';
+import '/util/platform_utils.dart';
+
+/// Service maintaining downloading and caching.
+class FileService {
+  /// Downloads file by the provided [url] and saves it to cache.
+  ///
+  /// Retries itself using exponential backoff algorithm on a failure.
+  static Future<Uint8List?> downloadAndCache(
+    String url,
+    String? checksum, {
+    Function(int count, int total)? onReceiveProgress,
+    CancelToken? cancelToken,
+    Future<void> Function()? onForbidden,
+  }) async {
+    File? file;
+    if (checksum != null) {
+      if (FIFOCache.exists(checksum)) {
+        return FIFOCache.get(checksum);
+      }
+
+      if (!PlatformUtils.isWeb) {
+        final Directory cache = await getApplicationDocumentsDirectory();
+        file = File('${cache.path}${Config.cache}/$checksum');
+        // TODO: remove before merge
+        // Android path - /data/user/0/com.team113.messenger/app_flutter/$Config.cache/$checksum
+        // Windows path - c:\Users\user\OneDrive\Документы\$Config.cache\$checksum
+
```

Comment:
@SleepySquash Добавил для Windows `Config.cache`, потому что напрямую в папку `Документы` наверное неправильно файлы сохранять. На Linux судя по [документации](https://wiki.archlinux.org/title/XDG_user_directories#:~:text=XDG_DOCUMENTS_DIR%3D%22%24HOME/Documents%22) ситуация аналогичная.

Для Android она не нужна, там сразу есть готовая папочка с названием приложения.

Посмотрите пожалуйста куда будут сохраняться файлики на iOS и macOS, возможно там тоже `Config.cache` не нужно использовать.

Reply:
@krida2000, [судя по документации `getApplicationDocumentsDirectory`](https://pub.dev/documentation/path_provider/latest/path_provider/getApplicationDocumentsDirectory.html): 

> Path to a directory where the application may place data that is user-generated, or that cannot otherwise be recreated by your application.

Эти файлики мы можем пересоздавать. Поэтому я не уверен, что директория корректная. В Windows же кэши лежат обычно в AppData'е, не в папке user'а.